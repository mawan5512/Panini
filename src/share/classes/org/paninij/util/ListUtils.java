/*
 * This file is part of the Panini project at Iowa State University.
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/.
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * For more details and the latest version of this code please see
 * http://paninij.org
 *
 * Contributor(s):
 */
package org.paninij.util;

import com.sun.tools.javac.util.List;

/**
 * Utility methods for operations on lists
 * @author Sean L. Mooney
 * @since panini-0.9.2
 */
public class ListUtils {

    private ListUtils() {}

    /**
     * Move the first element of the list that matches the predicate to
     * the front of the list.
     * @param capsule
     */
    public static <T> List<T> moveToFirst(List<T> capsule, Predicate<T> pred) {
       //nothing to reorder.
        if(capsule.size() < 2){
            return capsule;
        }
        if (pred.apply(capsule.head)) {
            return capsule;
        }

        List<T> head = capsule;
        List<T> curr = head.tail;
        List<T> prev = head;
        while(curr.nonEmpty()) {
            if (pred.apply(curr.head)) {
                prev.tail = curr.tail;
                curr.tail = head;
                head = curr;
                break;
            } else {
                prev = curr;
                curr = curr.tail;
            }
        }
        return head;
    }
}
