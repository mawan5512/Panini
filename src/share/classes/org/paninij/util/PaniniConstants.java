/*
 * This file is part of the Panini project at Iowa State University.
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/.
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 * 
 * For more details and the latest version of this code please see
 * http://paninij.org
 *
 * Contributor(s): Eric Lin
 */

package org.paninij.util;

public class PaniniConstants {
	public final static String PANINI = "Panini$";
	public final static String PANINI_QUEUE = "PaniniCapsule";
	public final static String PANINI_CAPSULE_SEQUENTIAL = "PaniniCapsuleSequential";
	public final static String PANINI_CAPSULE_THREAD = "PaniniCapsuleThread";
	public final static String PANINI_CAPSULE_TASK = "PaniniCapsuleTask";
	public final static String PANINI_CAPSULE_OBJECTS = "panini$capsule$objects";
	public final static String PANINI_CAPSULE_HEAD = "panini$capsule$head";
	public final static String PANINI_CAPSULE_TAIL = "panini$capsule$tail";
	public final static String PANINI_CAPSULE_SIZE = "panini$capsule$size";
	public final static String PANINI_CAPSULE_QUEUELOCK = "queueLock";
	public final static String PANINI_CAPSULE_EXTENDQUEUE = "panini$extendQueue";
	public final static String CAPSULE_CALL_QUEUES = "panini$msgNameQueue";
	public final static String CAPSULE_ARG_QUEUES = "panini$msgArgsQueue";
	public final static String CAPSULE_RETURN_QUEUES = "panini$msgDuckQueue";
	public final static String CAPSULE_METHOD_NAMES = "panini$methodConst";
	/**
	 * Name of the method used to wire an 'internal' system for a capsule.
	 */
	public final static String CAPSULE_SYS_WIRE = "panini$wire$sys";
	public final static String DUCK_INTERFACE_NAME = "Panini$Duck";
	public final static String PANINI_LAMBDA = "PaniniLambda";
	public final static String PANINI_LAMBDA_BODY = "lambda$body";
	public final static String ARRAY_DUCKS = "Array$Types";
	public final static String PANINI_FINISH = "panini$finish";
	public final static String PANINI_FINISH2 = "panini$finish2";//used for string lambdas
	public final static String PANINI_MESSAGE = "panini$message";
	public final static String PANINI_DUCK_GET = "panini$get";
	public final static String PANINI_GET_NEXT_DUCK = "get$Next$Duck";
	public final static String PANINI_TERMINATE = "panini$terminate";
	public final static String PANINI_METHOD_CONST = "panini$methodConst$";
	public final static String PANINI_MESSAGE_ID = "panini$message$id";
	public final static String PANINI_DUCK_TYPE = "panini$duck$future";
	public final static String PANINI_PUSH = "panini$push";
	public final static String PANINI_WRAPPED = "panini$wrapped";
	public final static String PANINI_SHUTDOWN = "shutdown";
	public final static String PANINI_YIELD = "yield";
	public final static String PANINI_EXIT = "exit";
	public final static String PANINI_START = "start";
	public final static String PANINI_JOIN = "join";
	public final static String PANINI_INIT = "panini$init";
	public final static String PANINI_CAPSULE_INIT = "panini$capsule$init";
	public final static String PANINI_WHEN = "panini$when";
	public final static String PANINI_CHECK_WHEN = "panini$check$when";
	public final static String PANINI_ORIGINAL_METHOD_SUFFIX = "$Original";
	public final static String PANINI_SUPERCAPSULE = "panini$superCapsule";

	public final static String PANINI_DUCK_HASHCODE = "hashCode";
	public final static String PANINI_DUCK_EQUALS = "equals";

	// To implement the flag in the duck.
	public final static String REDEEMED = "panini$redeemed";
	public final static String VALUE = "value";
	
	public final static String PANINI_REF_COUNT = "panini$ref$count";
	public final static String PANINI_DISCONNECT = "shutdown";

	public final static String UNRUNNABLE_EXCEPTION_CLASS =
	        "UnrunnableCapsuleException";
}
