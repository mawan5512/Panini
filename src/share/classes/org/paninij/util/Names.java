/*
 * This file is part of the Panini project at Iowa State University.
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/.
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * For more details and the latest version of this code please see
 * http://paninij.org
 *
 * Contributor(s):
 */
package org.paninij.util;

import com.sun.tools.javac.util.Name;
import org.paninij.util.PaniniConstants;

/**
 * Names for the panc compiler. Parallel to
 * {@link com.sun.tools.javac.util.Names}.
 * 
 * @author Sean L. Mooney
 * @since panini-0.9.2
 */
public class Names {
	/**
	 * Capsule Initialization method.
	 */
	public final Name PaniniCapsuleInit;
	/**
	 * Name of the 'run' method.
	 */
	public final Name Run;

	/** Name a start method. */
	public final Name Start;

	/**
	 * main
	 */
	public final Name Main;

	public final Name Capsule;
	/**
	 * Name used for a capsule (external) wiring symbol.
	 */
	public final Name Wiring;
	public final Name PaniniFinish;
	public final Name PaniniDuckFuture;
	public final Name PaniniLambdaBody;
	public final Name PaniniDisconnect;
	public final Name PaniniRefCountField;

	/**
	 * Name of the method used to wire internal 'systems' in a capsule.
	 */
	public final Name InternalCapsuleWiring;

	public final Name Org;
	public final Name Paninij;
	public final Name Runtime;
	public final Name UnrunnableCapsuleExceptionClass;

	/**
	 * Construct more names, using an exitings Names table.
	 * 
	 * @param names
	 */
	public Names(com.sun.tools.javac.util.Names names) {
		// Method Names
		PaniniCapsuleInit = names
				.fromString(PaniniConstants.PANINI_CAPSULE_INIT);
		Run = names.fromString("run");
		Start = names.fromString(PaniniConstants.PANINI_START);
		Main = names.fromString("main");

		// Capsule related
		Capsule = names.fromString("Capsule");
		Wiring = names.fromString("Wiring");
		PaniniFinish = names.fromString(PaniniConstants.PANINI_FINISH);
		PaniniLambdaBody = names.fromString(PaniniConstants.PANINI_LAMBDA_BODY);
		PaniniDuckFuture = names.fromString(PaniniConstants.PANINI_DUCK_TYPE);
		PaniniDisconnect = names.fromString(PaniniConstants.PANINI_DISCONNECT);
		InternalCapsuleWiring = names
				.fromString(PaniniConstants.CAPSULE_SYS_WIRE);
		PaniniRefCountField = names
				.fromString(PaniniConstants.PANINI_REF_COUNT);

		// Class and package Names
		Org = names.fromString("org");
		Paninij = names.fromString("paninij");
		Runtime = names.fromString("runtime");
		UnrunnableCapsuleExceptionClass = names
				.fromString(PaniniConstants.UNRUNNABLE_EXCEPTION_CLASS);
	}
}
