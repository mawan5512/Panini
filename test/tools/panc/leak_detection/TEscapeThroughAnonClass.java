/*
 * This file is part of the Panini project at Iowa State University.
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/.
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * For more details and the latest version of this code please see
 * http://paninij.org
 *
 * Contributor(s): Yuheng Long
 */

/*
 * @test
 * @summary Ensure Capsule state will not be used by other Capsule.
 * @compile/ref=TEscapeThroughAnonClass.out -XDrawDiagnostics TEscapeThroughAnonClass.java
 */

import java.util.ArrayList;

interface Escape {
	void method();
}

capsule B (TEscapeThroughAnonClass a) {
	ArrayList bal = new ArrayList();

	void escape(Escape fun){
		fun.method();
	}
}

capsule TEscapeThroughAnonClass {
	ArrayList aal = new ArrayList();
	int escapeei = 0;

	design {
		TEscapeThroughAnonClass a;
		B b;
		b(a);
	}

	void run() {
		b.escape(new Escape() {
			public void method() {
				escapeei = 2;
			}
		});
	}

	void nothing() {};
}
