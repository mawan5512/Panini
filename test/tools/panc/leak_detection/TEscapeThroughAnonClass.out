TEscapeThroughAnonClass.java:53:33: compiler.warn.confinement.violation: escapeei, TEscapeThroughAnonClass, run()
1 warning